File: Editor/src/main/java/com/rspsi/studio/theme/StudioIcons.java
package com.rspsi.studio.theme;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * Semantic icon registry for OpenRune Studio backed by Google Fonts Material Icons
 * (https://fonts.google.com/icons).
 *
 * Codepoints are mapped to the 16-bit BMP Private Use Area (0xE000 - 0xF8FF) and
 * rendered natively via Dear ImGui's font atlas.
 */
public final class StudioIcons {

    // --- Search & Discovery ---
    public static final String SEARCH = "\ue8b6";           // search
    public static final String EXPLORE = "\ue87a";          // explore
    public static final String NAVIGATION = "\ue55d";       // navigation
    public static final String SPEED = "\ue9e4";            // speed

    // --- Navigation, Viewports & Core Shell ---
    public static final String HOME = "\ue88a";             // home
    public static final String MAP = "\ue55b";              // map
    public static final String VIEWPORT = "\ue8f4";         // visibility (eye)
    public static final String VISIBILITY_OFF = "\ue8f5";   // visibility_off
    public static final String SETTINGS = "\ue8b8";         // settings (gear)
    public static final String SLIDERS = "\ue429";          // tune
    public static final String TUNE = "\ue429";             // tune
    public static final String INFO = "\ue88e";             // info
    public static final String HELP = "\ue887";             // help
    public static final String PIN = "\uf10d";              // push_pin
    public static final String CLOSE = "\ue5cd";            // close
    public static final String REFRESH = "\ue5d5";          // refresh
    public static final String FULLSCREEN = "\ue5d0";       // fullscreen
    public static final String OPEN_IN_NEW = "\ue89e";      // open_in_new

    // --- Selection & Editing Tools ---
    public static final String SELECT = "\ue569";           // near_me (pointer cursor)
    public static final String TILE = "\ue3c6";             // crop_square (single tile)
    public static final String AREA = "\ue162";             // select_all (marquee range)
    public static final String GRID = "\ue3ec";             // grid_on
    public static final String BRUSH = "\ue3ae";            // brush
    public static final String PALETTE = "\ue40a";          // palette
    public static final String COLOR_LENS = "\ue3b7";       // color_lens
    public static final String TEXTURE = "\ue421";          // texture
    public static final String LAYERS = "\ue53b";           // layers
    public static final String TERRAIN = "\ue564";          // terrain (mountain/landscape)
    public static final String HEIGHT = "\uea3b";           // architecture (elevation)
    public static final String STRAIGHTEN = "\ue41c";       // straighten
    public static final String TRENDING_UP = "\ue8e5";      // trending_up
    public static final String WATER = "\ue798";            // water_drop
    public static final String ENVIRONMENT = "\uea63";      // park (nature/environment)
    public static final String FLAG = "\ue153";             // flag (tile collision/mask)
    public static final String PATH = "\ue922";             // timeline (spline path)
    public static final String ROUTE = "\ueacd";            // route

    // --- 3D Objects & Hierarchy ---
    public static final String OBJECT = "\ue9fe";           // view_in_ar (3D cube)
    public static final String ADD_OBJECT = "\ue146";       // add_box
    public static final String PREFAB = "\ue1bd";           // widgets
    public static final String OUTLINER = "\ue8ef";         // view_list
    public static final String LIST = "\ue241";             // format_list_bulleted

    // --- Actions, History & File Operations ---
    public static final String HISTORY = "\ue889";          // history
    public static final String UNDO = "\ue166";             // undo
    public static final String REDO = "\ue15a";             // redo
    public static final String CHECK = "\ue5ca";            // check
    public static final String VALIDATE = "\ue5ca";         // check
    public static final String TERMINAL = "\ueb8e";         // terminal
    public static final String CODE = "\ue86f";             // code
    public static final String BUG_REPORT = "\ue868";       // bug_report
    public static final String FOLDER = "\ue2c7";           // folder
    public static final String FOLDER_OPEN = "\ue2c8";      // folder_open
    public static final String SAVE = "\ue161";             // save
    public static final String DELETE = "\ue872";           // delete
    public static final String EDIT = "\ue3c9";             // edit
    public static final String COPY = "\ue14d";             // content_copy
    public static final String PASTE = "\ue14f";            // content_paste
    public static final String PLAY = "\ue037";             // play_arrow

    // --- Chevrons & UI Directionals ---
    public static final String EXPAND_MORE = "\ue5cf";      // expand_more (down chevron)
    public static final String EXPAND_LESS = "\ue5ce";      // expand_less (up chevron)
    public static final String CHEVRON_LEFT = "\ue5cb";     // chevron_left
    public static final String CHEVRON_RIGHT = "\ue5cc";    // chevron_right
    public static final String ARROW_DROP_DOWN = "\ue5c5";  // arrow_drop_down
    public static final String ARROW_DROP_UP = "\ue5c7";    // arrow_drop_up

    private static final Map<String, String> ICONS_BY_NAME;

    static {
        Map<String, String> map = new HashMap<>(2500);
        try (InputStream is = StudioIcons.class.getResourceAsStream("/font/MaterialIcons-Regular.codepoints")) {
            if (is != null) {
                try (BufferedReader reader = new BufferedReader(new InputStreamReader(is, StandardCharsets.UTF_8))) {
                    String line;
                    while ((line = reader.readLine()) != null) {
                        int space = line.indexOf(' ');
                        if (space > 0) {
                            String name = line.substring(0, space).trim();
                            String hex = line.substring(space + 1).trim();
                            try {
                                int cp = Integer.parseInt(hex, 16);
                                if (cp <= 0xFFFF) {
                                    map.put(name, String.valueOf((char) cp));
                                }
                            } catch (NumberFormatException ignored) {}
                        }
                    }
                }
            }
        } catch (Exception ignored) {}
        ICONS_BY_NAME = Collections.unmodifiableMap(map);
    }

    private StudioIcons() {
    }

    /**
     * Looks up a Google Fonts Material Icon glyph by its standard name
     * (e.g. "brush", "water_drop", "sports_esports", "layers").
     *
     * @param name The Google Fonts icon name.
     * @return The icon glyph string, or fallback if not found.
     */
    public static String byName(String name, String fallback) {
        if (name == null || name.isBlank()) return fallback;
        return ICONS_BY_NAME.getOrDefault(name.toLowerCase().trim(), fallback);
    }

    /**
     * Looks up a Google Fonts Material Icon glyph by its standard name,
     * defaulting to OBJECT ("view_in_ar") if not found.
     */
    public static String byName(String name) {
        return byName(name, OBJECT);
    }

    /**
     * Returns true if the Google Fonts Material Icons registry contains the given icon name.
     */
    public static boolean has(String name) {
        return name != null && ICONS_BY_NAME.containsKey(name.toLowerCase().trim());
    }

    /**
     * Total number of registered Google Material Icons.
     */
    public static int registeredIconCount() {
        return ICONS_BY_NAME.size();
    }
}

File: Editor/src/main/kotlin/com/rspsi/studio/theme/StudioIcons.kt
package com.rspsi.studio.theme

import java.nio.charset.StandardCharsets
import java.util.Locale

/**
 * Semantic icon registry for OpenRune Studio backed by Google Material Icons.
 *
 * Named constants are the stable application vocabulary. [byName] additionally exposes
 * the bundled Material Icons codepoint catalog for extension/tool metadata that provides
 * icon names dynamically at runtime.
 */
object StudioIcons {
    // Search & discovery
    const val SEARCH = "\ue8b6"
    const val EXPLORE = "\ue87a"
    const val NAVIGATION = "\ue55d"
    const val SPEED = "\ue9e4"

    // Navigation, viewports & shell
    const val HOME = "\ue88a"
    const val MAP = "\ue55b"
    const val VIEWPORT = "\ue8f4"
    const val VISIBILITY_OFF = "\ue8f5"
    const val SETTINGS = "\ue8b8"
    const val EXTENSION = "\ue87b"
    const val SLIDERS = "\ue429"
    const val TUNE = "\ue429"
    const val INFO = "\ue88e"
    const val PLAYER = "\ue7fd"
    const val HELP = "\ue887"
    const val PIN = "\uf10d"
    const val CLOSE = "\ue5cd"
    const val REFRESH = "\ue5d5"
    const val FULLSCREEN = "\ue5d0"
    const val FULLSCREEN_EXIT = "\ue5d1"
    const val MINIMIZE = "\ue931"
    const val OPEN_IN_NEW = "\ue89e"
    const val LOCK = "\ue897"
    const val LOCK_OPEN = "\ue898"
    const val MORE_VERT = "\ue5d4"
    const val MORE_HORIZ = "\ue5d3"
    const val CAMERA = "\ue412"
    const val CENTER_FOCUS = "\ue3dc"
    const val ZOOM_IN = "\ue8ff"
    const val ZOOM_OUT = "\ue900"

    // Selection & editing
    const val SELECT = "\ue569"
    const val TILE = "\ue3c6"
    const val AREA = "\ue162"
    const val GRID = "\ue3ec"
    const val GRID_OFF = "\ue3eb"
    const val BRUSH = "\ue3ae"
    const val PALETTE = "\ue40a"
    const val COLOR_LENS = "\ue3b7"
    const val TEXTURE = "\ue421"
    const val LAYERS = "\ue53b"
    const val LAYERS_CLEAR = "\ue53c"
    const val TERRAIN = "\ue564"
    const val HEIGHT = "\uea3b"
    const val STRAIGHTEN = "\ue41c"
    const val TRENDING_UP = "\ue8e5"
    const val WATER = "\ue798"
    const val ENVIRONMENT = "\uea63"
    const val FLAG = "\ue153"
    const val PATH = "\ue922"
    const val ROUTE = "\ueacd"
    const val POLYLINE = "\uebbb"

    // Brush shapes & geometry
    const val SHAPE_SQUARE = "\ue3c6"
    const val SHAPE_CIRCLE = "\uef4a"
    const val SHAPE_DIAMOND = "\uead5"
    const val SHAPE_FALLOFF = "\ue3e9"
    const val SHAPE_WAVE = "\ue176"

    // Height sculpting
    const val RAISE = "\ue5d8"
    const val LOWER = "\ue5db"
    const val FLATTEN = "\uf108"
    const val SMOOTH = "\ue176"
    const val TERRACE = "\uf1a9"
    const val BLEND = "\ue429"

    // Steppers & transforms
    const val ADD = "\ue145"
    const val REMOVE = "\ue15b"
    const val ROTATE_LEFT = "\ue419"
    const val ROTATE_RIGHT = "\ue41a"

    // 3D objects & hierarchy
    const val OBJECT = "\ue9fe"
    const val ADD_OBJECT = "\ue146"
    const val PREFAB = "\ue1bd"
    const val OUTLINER = "\ue8ef"
    const val LIST = "\ue241"

    // Actions, history & files
    const val HISTORY = "\ue889"
    const val UNDO = "\ue166"
    const val REDO = "\ue15a"
    const val CHECK = "\ue5ca"
    const val VALIDATE = "\ue5ca"
    const val TERMINAL = "\ueb8e"
    const val CODE = "\ue86f"
    const val BUG_REPORT = "\ue868"
    const val FOLDER = "\ue2c7"
    const val FOLDER_OPEN = "\ue2c8"
    const val SAVE = "\ue161"
    const val DELETE = "\ue872"
    const val EDIT = "\ue3c9"
    const val COPY = "\ue14d"
    const val PASTE = "\ue14f"
    const val PLAY = "\ue037"
    const val PAUSE = "\ue034"
    const val STOP = "\ue047"

    // Chevrons & directionals
    const val EXPAND_MORE = "\ue5cf"
    const val EXPAND_LESS = "\ue5ce"
    const val CHEVRON_LEFT = "\ue5cb"
    const val CHEVRON_RIGHT = "\ue5cc"
    const val ARROW_DROP_DOWN = "\ue5c5"
    const val ARROW_DROP_UP = "\ue5c7"

    private val ICONS_BY_NAME: Map<String, String> = loadNamedIcons()

    /**
     * Resolves a Google Material Icon name such as "brush" or "water_drop".
     *
     * Dynamic lookup is intentionally best-effort: if the bundled codepoint catalog cannot
     * be read, the stable constants above still work and callers receive their fallback.
     */
    @JvmStatic
    fun byName(
        name: String?,
        fallback: String,
    ): String {
        val key = normalizeName(name) ?: return fallback
        return ICONS_BY_NAME[key] ?: fallback
    }

    /** Resolves a name and falls back to [OBJECT] when it is unknown. */
    @JvmStatic
    fun byName(name: String?): String = byName(name, OBJECT)

    @JvmStatic
    fun has(name: String?): Boolean {
        val key = normalizeName(name) ?: return false
        return ICONS_BY_NAME.containsKey(key)
    }

    @JvmStatic
    fun registeredIconCount(): Int = ICONS_BY_NAME.size

    private fun normalizeName(name: String?): String? =
        name
            ?.trim()
            ?.takeIf { it.isNotEmpty() }
            ?.lowercase(Locale.ROOT)

    private fun loadNamedIcons(): Map<String, String> {
        val input =
            StudioIcons::class.java.getResourceAsStream(
                "/font/MaterialIcons-Regular.codepoints",
            ) ?: return emptyMap()

        // This registry is supplemental metadata, not startup-critical UI state. Malformed
        // lines are skipped so one catalog entry cannot disable all icon-name resolution.
        return runCatching {
            input.bufferedReader(StandardCharsets.UTF_8).useLines { lines ->
                buildMap(capacity = 2500) {
                    lines.forEach { line ->
                        val separator = line.indexOf(' ')
                        if (separator <= 0) {
                            return@forEach
                        }

                        val name = line.substring(0, separator).trim()
                        val hex = line.substring(separator + 1).trim()
                        val codePoint = hex.toIntOrNull(16) ?: return@forEach
                        if (codePoint <= 0xFFFF) {
                            put(name, codePoint.toChar().toString())
                        }
                    }
                }
            }
        }.getOrDefault(emptyMap())
    }
}
